elfloader-tool/src/common.c

```diff
@@ -369,16 +374,21 @@ int load_images(
     size_t cpio_len = _archive_start_end - _archive_start;
 
     /* Load kernel. */
-    unsigned long kernel_elf_blob_size;
+    unsigned long cpio_file_size = 0;
     void const *kernel_elf_blob = cpio_get_file(cpio,
                                                 cpio_len,
                                                 "kernel.elf",
-                                                &kernel_elf_blob_size);
+                                                &cpio_file_size);
     if (kernel_elf_blob == NULL) {
         printf("ERROR: No kernel image present in archive\n");
         return -1;
     }
 
+    /* Ensure we can safely cast the CPIO API type to our preferred type. */
+    _Static_assert(sizeof(cpio_file_size) <= sizeof(size_t),
```

Comment:
Ah, how did you track this down?  I noticed that after earlier changes introduced that reordered some of the parameters to functions in this file, the system would fault when run at higher optimization levels. So the issue turns out to be due to libcpio being compiled under different integer size?

Reply:
Actually, this was manual code review. After the resent 32/64 bit surprises I get a bit uneasy every time a "long" shows up. So I started to add safeguards like this to code that works with integers from APIs with different integer models. I follow the simple rule of thumb (a) "better safe than sorry" and (b) avoid using `long`and let it spread.
Candidate high in the list were the CPIO API and the ELF lib API, because they face the fundamental problem of serializing/deserializing data into storage formats that are architecture agnostic. Turned out adding the compile time assert comes handy. No need to look deeper in to this as long as it does not hit. And also no need to write any runtime code that tries to handle conversion corner cases. 
So far these asserts don't hit, so everything is fine. If things ever change, there will be a build breaker.

Reply:
Right.  I'm currently testing some changes that remove the CPIO archive and directly link the files into the binary under named symbol_start and symbol_end [1].  Given that the elfloader only loads specific files and requires them to have specific names, it doesn't seem like the CPIO is giving us much benefit.  Additionally not using a CPIO allows us to set the alignment of the files which can make loading them more efficient.
[1]:
```
.globl ${symbol}, ${symbol}_end
.balign 0x1000
${symbol}:
.incbin \"${CMAKE_CURRENT_BINARY_DIR}/${symbol}\"
${symbol}_end:
```

Reply:
Frankly, I've been wondering what the elfloader plans are actually. It seem to be prepared do to much more than what it is currently doing. For example the comments say that is can load different ELF to different cores - but this is not used and the core actually check that is loads exactly one image. Nevertheless is is a handy tool and using CPIO keeps it flexible.  There is this hash check feature that adds more files in the CPIO and this really works. So when you remove CPIO now, how will that work?
But I agree, if the images are better aligned, then copying around the segments would not longer be necessary. In the end, we don't even need to keep ELF files in there, but use a more streamlines load format with bianry images and just a litte bit of meta data. But this road leads to a much more specific loader for dedicated usage. I was wondering if the ELF loader is just a generic nice swiss army knife tool for developers.